core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,250 @@
+
+  /**
+   * Checks to see if there is enough memory on the cluster to hold the cached blocks on these
+   * executors. Checks on a per executor basis from the smallest to the largest.
+   *
+   * @param execIds the executors to query
+   * @return a map of executor ids to whether or not there is enough free memory for them.
+   */
+  private def checkFreeMem(execIds: Seq[String]): Map[String, Boolean] = {
+    val execIdsToShutDown = execIds.toSet
+    val allExecMemStatus: Map[String, (Long, Long)] = blockManagerMasterEndpoint
+      .askSync[Map[BlockManagerId, (Long, Long)]](GetMemoryStatus)
+      .map { case (blockManagerId, mem) => blockManagerId.executorId -> mem }
+
+    val (expiringMemStatus, remainingMemStatus) = allExecMemStatus.partition {
+      case (k, v) => execIdsToShutDown.contains(k)
+    }
+    val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+    val alreadyReplicatedBlocks: Map[String, Set[RDDBlockId]] =
+      savedBlocks.filterKeys(execIdsToShutDown).map { case (k, v) => k -> v.toSet }.toMap
+
+    val getSizes = GetSizeOfBlocks(alreadyReplicatedBlocks)
+    val alreadyReplicatedBytes = blockManagerMasterEndpoint.askSync[Map[String, Long]](getSizes)
+
+    val bytesToReplicate: Seq[(String, Long)] =
+      expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+        val alreadyReplicated = alreadyReplicatedBytes.getOrElse(execId, 0L)
+        val usedMem = maxMem - remainingMem
+        val toBeReplicatedMem = usedMem - alreadyReplicated
+        execId -> toBeReplicatedMem
+      }.toSeq.sortBy { case (k, v) => v }
+
+    bytesToReplicate.scan(("start", freeMemOnRemaining)) {
+      case ((_, remainingMem), (execId, mem)) => (execId, remainingMem - mem)
+    }.drop(1)
+      .toMap
+      .mapValues(freeMem => freeMem >= 0)
+  }
+
+  /**
+   * Gets a list of cached blocks from the block manager master and updates.
+   * CacheRecoveryManagerState.
+   *
+   * @param execId an executor Id
+   */
+  private def updateBlockState(execId: String): Unit = {
+    val blocks: mutable.Set[RDDBlockId] = blockManagerMasterEndpoint
+      .askSync[collection.Set[BlockId]](GetCachedBlocks(execId))
+      .flatMap(_.asRDDId)(collection.breakOut)
+
+    blocks --= savedBlocks.getOrElse(execId, new mutable.HashSet)
+    blocksToSave(execId) = mutable.PriorityQueue[RDDBlockId](blocks.toSeq: _*)(Ordering.by(_.rddId))
+    killTimers.getOrElseUpdate(execId, scheduleKill(new KillRunner(execId)))
+  }
+
+  class KillRunner(execId: String) extends Runnable { def run(): Unit = killExecutor(execId) }
+  private def scheduleKill(k: KillRunner) = scheduler.schedule(k, forceKillAfterS, TimeUnit.SECONDS)
+
+  /**
+   * Ask block manager master to replicate one cached block.
+   *
+   * @param execId the executor to replicate a block from
+   * @return false if there are no more blocks or if there is an issue
+   */
+  def replicateFirstBlock(execId: String): (Future[Boolean], Option[RDDBlockId]) = synchronized {
```

Comment:
This method is only called in one place. It may be simpler to just inline it there so you don't have to explain this odd return type.

Reply:
this return type is simpler now.